sway-core/src/type_system/id.rs

```diff
@@ -483,48 +492,115 @@ impl TypeId {
+                                    trait_name: format!(
+                                        "{}{}",
+                                        structure_trait_constraint.trait_name.suffix,
+                                        type_arguments_string
+                                    ),
                                     span: span.clone(),
                                 });
-                            }
-                        }
+                            },
+                        );
                     }
                 }
             }
             Ok(())
         })
     }
+
+    fn check_trait_constraints_errors(
+        &self,
+        handler: &Handler,
+        mut ctx: TypeCheckContext,
+        structure_type_id: &TypeId,
+        structure_trait_constraints: &Vec<TraitConstraint>,
+        f: impl Fn(&TraitConstraint),
+    ) -> bool {
+        let engines = ctx.engines();
+        let unify_check = UnifyCheck::non_dynamic_equality(engines);
+        let mut found_error = false;
+        let generic_trait_constraints_trait_names_and_args = ctx
+            .namespace
+            .implemented_traits
+            .get_trait_names_and_type_arguments_for_type(engines, *structure_type_id);
+        for structure_trait_constraint in structure_trait_constraints {
+            let structure_trait_constraint_trait_name = &structure_trait_constraint
+                .trait_name
+                .to_fullpath(ctx.namespace);
+            if !generic_trait_constraints_trait_names_and_args.iter().any(
+                |(trait_name, trait_args)| {
+                    trait_name == structure_trait_constraint_trait_name
+                        && trait_args.len() == structure_trait_constraint.type_arguments.len()
+                        && trait_args
+                            .iter()
+                            .zip(structure_trait_constraint.type_arguments.iter())
+                            .all(|(t1, t2)| {
+                                unify_check.check(
+                                    self.check_trait_constraints_errors_resolve(
+                                        handler,
+                                        ctx.by_ref(),
+                                        t1.type_id,
+                                        t1.span.clone(),
+                                    ),
+                                    self.check_trait_constraints_errors_resolve(
+                                        handler,
+                                        ctx.by_ref(),
+                                        t2.type_id,
+                                        t2.span.clone(),
+                                    ),
+                                )
+                            })
+                },
+            ) {
+                found_error = true;
+                f(structure_trait_constraint);
+            }
+        }
+        found_error
+    }
+
+    fn check_trait_constraints_errors_resolve(
```

Comment:
My strong preference here would be making the `check_trait_constraints_errors_resolve` a nested function inside of the `check_trait_constraints_errors_resolve` and naming it e.g. `resolve_type`. Why? It is used only within the `check_trait_constraints_errors_resolve` and putting it to the `TypeId` level bloats IMO the private interface of `TypeId`. Also, I think the method actually do not need `&self`.